include/zephyr/dfu/mcuboot_partitions.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DFU_BOOT_MCUBOOT_PARTITIONS_H_
+#define ZEPHYR_DFU_BOOT_MCUBOOT_PARTITIONS_H_
+
+#include <zephyr/storage/flash_map.h>
+
+#if (defined(CONFIG_UPDATEABLE_IMAGE_NUMBER) && CONFIG_UPDATEABLE_IMAGE_NUMBER > 1)
+#define ZEPHYR_MCUBOOT_APP_IMAGE_COUNT CONFIG_UPDATEABLE_IMAGE_NUMBER
+#else
+#define ZEPHYR_MCUBOOT_APP_IMAGE_COUNT 1
+#endif
+
+#ifdef CONFIG_TRUSTED_EXECUTION_NONSECURE
+#define ZEPHYR_MCUBOOT_APP_0_PRIMARY_SLOT_LABEL		slot0_ns_partition
+#define ZEPHYR_MCUBOOT_APP_0_SECONDARY_SLOT_LABEL	slot1_ns_partition
+#else
+#define ZEPHYR_MCUBOOT_APP_0_PRIMARY_SLOT_LABEL		slot0_partition
+#define ZEPHYR_MCUBOOT_APP_0_SECONDARY_SLOT_LABEL	slot1_partition
+#endif /* CONFIG_TRUSTED_EXECUTION_NONSECURE */
+
+#define ZEPHYR_MCUBOOT_APP_1_PRIMARY_SLOT_LABEL		slot2_partition
+#define ZEPHYR_MCUBOOT_APP_1_SECONDARY_SLOT_LABEL	slot3_partition
+
+#define ZEPHYR_MCUBOOT_APP_0_PRIMARY_SLOT_EXISTS \
+	FIXED_PARTITION_EXISTS(ZEPHYR_MCUBOOT_APP_0_PRIMARY_SLOT_LABEL)
+#define ZEPHYR_MCUBOOT_APP_0_SECONDARY_SLOT_EXISTS \
+	FIXED_PARTITION_EXISTS(ZEPHYR_MCUBOOT_APP_0_SECONDARY_SLOT_LABEL)
+#define ZEPHYR_MCUBOOT_APP_1_PRIMARY_SLOT_EXISTS \
+	FIXED_PARTITION_EXISTS(ZEPHYR_MCUBOOT_APP_1_PRIMARY_SLOT_LABEL)
+#define ZEPHYR_MCUBOOT_APP_1_SECONDARY_SLOT_EXISTS \
+	FIXED_PARTITION_EXISTS(ZEPHYR_MCUBOOT_APP_1_SECONDARY_SLOT_LABEL)
+
+#if ZEPHYR_MCUBOOT_APP_0_SECONDARY_SLOT_EXISTS
+#define ZEPHYR_MCUBOOT_APP_IS_SINGLE_SLOT 0
+#else
+#define ZEPHYR_MCUBOOT_APP_IS_SINGLE_SLOT 1
+#endif
```

Comment:
Would need to follow how this is used, but this should only allow the option of non-single slot mode being used and not force anything, a person might have 2 slots but only want to use single slot mode (i.e. secondary slot could be a permanent fallback image which is entered somehow)

Reply:
This comes from MCUboot configuration where you can configure MCUboot to work with single slot and application that can not update itself.
The MCUboot has a Kconfig option `CONFIG_SINGLE_APPLICATION_SLOT` and no restrictions on existence of slots, except that slot0 needs to exist.
In case of Zephyr, the logic is implicitly assuming singe slot operation, looking at the mcuboot shell or the USB DFU, by the existence of the secondary slot (slot1). Actually the USB DFU in MCUboot can upload both slots if they exist and in app it can exist the opposite from the one it runs from, so when there is only slot0, then USB DFU will only work in MCUboot.
Maybe we should have Kconfig, somewhere around `CONFIG_MCUBOOT_BOOTLOADER` , equivalent of `CONFIG_SINGLE_APPLICATION_SLOT` in MCUboot, that could be set for non`CONFIG_MCUBOOT` app builds to set app as non-upgradeable (from itself) application. With such Kconfig we could have the logic reversed, where we would expect two slots to exist when the option is not set.